presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java

```diff
@@ -50,15 +60,22 @@ public static Block distinct(@TypeParameter("E") Type type, @SqlType("array(E)")
         }
 
         TypedSet typedSet = new TypedSet(type, array.getPositionCount(), "array_distinct");
-        BlockBuilder distinctElementBlockBuilder = type.createBlockBuilder(null, array.getPositionCount());
+        BlockBuilder distinctElementBlockBuilder = pageBuilder.getBlockBuilder(0);
         for (int i = 0; i < array.getPositionCount(); i++) {
             if (!typedSet.contains(array, i)) {
                 typedSet.add(array, i);
+                length++;
                 type.appendTo(array, i, distinctElementBlockBuilder);
             }
         }
 
-        return distinctElementBlockBuilder.build();
+        if (pageBuilder.isFull()) {
+            pageBuilder.reset();
+        }
+
+        pageBuilder.declarePositions(length);
+
+        return distinctElementBlockBuilder.getRegion(distinctElementBlockBuilder.getPositionCount() - length, length);
     }
 
     @SqlType("array(bigint)")
```

Comment:
We want to make this BIGINT specialized implementation to also use PageBuilder. 

However, I realized it's not straightforward, since this doesn't include TypeParameter `E`. And scalar function engine cannot  construct the PageBuilder.

One way is to make `bigintDistinct` specialized on on native container type instead of a specific Presto type.  Since the current implementation is tied to a `LongSet`,  we will switch to use a `TypedSet` as well.

Thoughts @haozhun ?